core/TypePtr.h

```diff
@@ -4,25 +4,148 @@
 #include <memory>
 
 namespace sorbet::core {
-// using TypePtr = std::shared_ptr<Type>;
 class Type;
-class TypePtr {
-    std::shared_ptr<Type> store;
-    TypePtr(std::shared_ptr<Type> &&store);
+class TypePtr final {
+public:
+    // We store tagged pointers as 64-bit values.
+    using tagged_storage = u8;
+
+    enum class Tag {
+        ClassType = 1,
+        LambdaParam,
+        SelfTypeParam,
+        AliasType,
+        SelfType,
+        LiteralType,
+        TypeVar,
+        OrType,
+        AndType,
+        ShapeType,
+        TupleType,
+        AppliedType,
+        MetaType,
+        BlamedUntyped,
+        UnresolvedClassType,
+        UnresolvedAppliedType,
+    };
+
+    // A mapping from type to its corresponding tag.
+    template <typename T> struct TypeToTag;
+
+private:
+    std::atomic<u4> *counter;
```

Comment:
Instead of allocating a counter, can we just pack the "has counter" flag and the counter itself into an 8-byte field, which is the same size as the pointer here, but saves an indirection and allocation?

Reply:
@froydnj how would I share the counter with other equivalent TypePtr instances without indirection?

Reply:
Ah, that wouldn't work very well, would it?  Never mind me!

Reply:
No worries; I actually had the same thought for a hot second during development before realizing the issue. :D